File: fetch.js
const axios = require("axios");
const fs = require("fs");

const words = [];

(async () => {
  let x = await axios.get(
    `https://zh.moegirl.org/api.php?action=query&list=allpages&format=json&aplimit=max`
  );
  while (1) {
    words.push(...x.data.query.allpages.map((x) => x.title));
    console.log(`Got ${words.length} entries`);
    fs.writeFileSync("titles.txt", words.join("\n"));
    if (x.data.continue) {
      console.log("Next:", x.data.continue.apcontinue);
      x = await axios.get(
        `https://zh.moegirl.org/api.php?action=query&list=allpages&format=json&aplimit=max&apcontinue=${encodeURIComponent(
          x.data.continue.apcontinue
        )}`
      );
    } else {
      break;
    }
  }
})();

File: fetch.py
#!/usr/bin/env python
import sys
import json
from urllib.request import Request, urlopen
from urllib.parse import quote_plus
from retry import retry


class StatusError(Exception):
    def __init__(self, code):
        super().__init__("HTTP status is {}".format(code))


@retry()
def open_request(url):
    return urlopen(Request(url, headers={
        "User-Agent": "PyFcitxDictBot/1.0; github.com/outloudvi/fcitx5-pinyin-moegirl"
    }))

def fetch(url):
    res = open_request(url)
    if res.status == 200:
        return json.loads(res.read())
    else:
        raise StatusError(res.status)


API_PAGE = sys.argv[1]
FILE = "titles.txt"
if len(sys.argv) > 2:
    FILE = sys.argv[2]
articles = []
data = fetch(API_PAGE + "?action=query&list=allpages&format=json")
while True:
    for i in map(lambda x: x["title"], data["query"]["allpages"]):
        articles.append(i)
    print("Got {} pages".format(len(articles)))
    if "continue" in data:
        data = fetch(API_PAGE + "?action=query&list=allpages&format=json&aplimit=max&apcontinue={}".format(
            quote_plus(data["continue"]["apcontinue"])))
    else:
        break
print("Finished.")
with open(FILE, "w") as f:
    f.write("\n".join(articles))
